File: 006_zigzag_conversion/Solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1: return s

        rows = [''] * numRows
        cur_row = 0
        downward = False
        for ch in s:
            rows[cur_row] += ch
            if cur_row in [0, numRows - 1]: downward = not downward
            cur_row += 1 if downward else -1

        return ''.join(rows)

    def convert2(self, s: str, numRows: int) -> str:
        if numRows == 1: return s

        rows = [[] for i in range(numRows)]
        cur_row = 0
        downward = False
        for ch in s:
            rows[cur_row].append(ch)
            if cur_row in [0, numRows - 1]: downward = not downward
            cur_row += 1 if downward else -1
        return ''.join([''.join(x) for x in rows])


if __name__ == '__main__':
    solution = Solution()
    print(solution.convert2('LEETCODEISHIRING', 3))

File: 006_zigzag_conversion/Solution.go
package main

import (
    "fmt"
    "strings"
)

func convert(s string, numRows int) string {
    if numRows == 1 {
        return s
    }

    rows := make([]string, numRows)
    curRow := 0
    downward := false

    for _, v := range s {
        rows[curRow] += string(v)
        if curRow == 0 || curRow == numRows-1 {
            downward = !downward
        }
        if downward {
            curRow += 1
        } else {
            curRow -= 1
        }
    }

    return strings.Join(rows, "")
}

func main() {
    fmt.Println(convert("LEETCODEISHIRING", 3))
}
